File: client/src/lib/geo.ts
/** Schlanke Geo-Utils — equirektangulare Näherung reicht für <2 km Distanzen. */

export interface LngLat {
  lng: number;
  lat: number;
}

const R = 6371000;
const DEG = Math.PI / 180;

/** Meter-Distanz zweier Punkte (equirektangular). */
export function distanceM(a: LngLat, b: LngLat): number {
  const x = (b.lng - a.lng) * DEG * Math.cos(((a.lat + b.lat) / 2) * DEG);
  const y = (b.lat - a.lat) * DEG;
  return Math.sqrt(x * x + y * y) * R;
}

/** de-DE-Distanz: unter 1 km "650 m", darüber "2,1 km". */
export function formatDistanceM(m: number): string {
  return m >= 1000 ? (m / 1000).toFixed(1).replace(".", ",") + " km" : Math.round(m) + " m";
}

/** Punkt-in-Ring (Ray-Casting), Ring = [[lng,lat],...]. */
function inRing(p: LngLat, ring: number[][]): boolean {
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const [xi, yi] = ring[i];
    const [xj, yj] = ring[j];
    if (yi > p.lat !== yj > p.lat && p.lng < ((xj - xi) * (p.lat - yi)) / (yj - yi) + xi) {
      inside = !inside;
    }
  }
  return inside;
}

/** Punkt in Polygon (äußerer Ring minus Löcher). coordinates = Polygon-Koordinaten. */
export function pointInPolygon(p: LngLat, coordinates: number[][][]): boolean {
  if (!coordinates.length || !inRing(p, coordinates[0])) return false;
  for (let i = 1; i < coordinates.length; i++) {
    if (inRing(p, coordinates[i])) return false;
  }
  return true;
}

/** Kürzeste Meter-Distanz Punkt → Segment (in lokaler Meter-Projektion). */
function distToSegmentM(p: LngLat, a: number[], b: number[]): number {
  const cos = Math.cos(p.lat * DEG);
  const ax = (a[0] - p.lng) * DEG * cos * R;
  const ay = (a[1] - p.lat) * DEG * R;
  const bx = (b[0] - p.lng) * DEG * cos * R;
  const by = (b[1] - p.lat) * DEG * R;
  const dx = bx - ax;
  const dy = by - ay;
  const len2 = dx * dx + dy * dy;
  const t = len2 === 0 ? 0 : Math.max(0, Math.min(1, -(ax * dx + ay * dy) / len2));
  const cx = ax + t * dx;
  const cy = ay + t * dy;
  return Math.sqrt(cx * cx + cy * cy);
}

/** Kürzeste Meter-Distanz Punkt → Polygonkante (alle Ringe). */
export function distToPolygonEdgeM(p: LngLat, coordinates: number[][][]): number {
  let min = Infinity;
  for (const ring of coordinates) {
    for (let i = 0; i < ring.length - 1; i++) {
      const d = distToSegmentM(p, ring[i], ring[i + 1]);
      if (d < min) min = d;
    }
  }
  return min;
}

File: ios/Packages/GreenZonesKit/Sources/GreenZonesKit/Geo/Geo.swift
import CoreLocation
import Foundation

/// Port von `client/src/lib/geo.ts` — equirektangulare Naeherung, reicht fuer
/// Distanzen < 2 km. Die Formeln sind absichtlich Zeichen fuer Zeichen dieselben
/// wie in v1: der Vektor-Test vergleicht auf 1,5 m genau, jede „Verbesserung"
/// waere eine Abweichung.

/// Ein Ring liegt flach als `[lng, lat, lng, lat, …]` im Speicher — ein
/// `[[Double]]` je Punkt kostet in Hannover-Tiles fuenfstellig viele
/// Array-Allokationen pro Kachel.
public typealias GZRing = [Double]
/// Polygon: Ring 0 ist die Aussenkante, alle weiteren sind Loecher.
public typealias GZPolygon = [GZRing]

public enum Geo {
    static let earthRadiusM = 6_371_000.0
    static let deg = Double.pi / 180

    /// Meter-Distanz zweier Punkte (equirektangular).
    public static func distanceM(_ a: CLLocationCoordinate2D, _ b: CLLocationCoordinate2D) -> Double {
        let x = (b.longitude - a.longitude) * deg * cos((a.latitude + b.latitude) / 2 * deg)
        let y = (b.latitude - a.latitude) * deg
        return (x * x + y * y).squareRoot() * earthRadiusM
    }

    /// de-DE-Distanz: unter 1 km „650 m", darueber „2,1 km".
    public static func formatDistanceM(_ m: Double) -> String {
        if m >= 1000 {
            // `toFixed(1)` rundet kaufmaennisch — `String(format:)` tut dasselbe.
            let km = String(format: "%.1f", m / 1000).replacingOccurrences(of: ".", with: ",")
            return km + " km"
        }
        return "\(Int(m.rounded())) m"
    }

    /// Punkt-in-Ring (Ray-Casting).
    static func inRing(_ p: CLLocationCoordinate2D, _ ring: GZRing) -> Bool {
        var inside = false
        let count = ring.count / 2
        guard count > 0 else { return false }
        var j = count - 1
        for i in 0..<count {
            let xi = ring[2 * i], yi = ring[2 * i + 1]
            let xj = ring[2 * j], yj = ring[2 * j + 1]
            if (yi > p.latitude) != (yj > p.latitude),
               p.longitude < (xj - xi) * (p.latitude - yi) / (yj - yi) + xi {
                inside.toggle()
            }
            j = i
        }
        return inside
    }

    /// Punkt in Polygon (aeusserer Ring minus Loecher).
    public static func pointInPolygon(_ p: CLLocationCoordinate2D, _ polygon: GZPolygon) -> Bool {
        guard let outer = polygon.first, inRing(p, outer) else { return false }
        for i in 1..<max(polygon.count, 1) where inRing(p, polygon[i]) {
            return false
        }
        return true
    }

    /// Kuerzeste Meter-Distanz Punkt → Segment (in lokaler Meter-Projektion).
    static func distToSegmentM(px: Double, py: Double, cosLat: Double,
                               ax0: Double, ay0: Double, bx0: Double, by0: Double) -> Double {
        let ax = (ax0 - px) * deg * cosLat * earthRadiusM
        let ay = (ay0 - py) * deg * earthRadiusM
        let bx = (bx0 - px) * deg * cosLat * earthRadiusM
        let by = (by0 - py) * deg * earthRadiusM
        let dx = bx - ax
        let dy = by - ay
        let len2 = dx * dx + dy * dy
        let t = len2 == 0 ? 0 : max(0, min(1, -(ax * dx + ay * dy) / len2))
        let cx = ax + t * dx
        let cy = ay + t * dy
        return (cx * cx + cy * cy).squareRoot()
    }

    /// Kuerzeste Meter-Distanz Punkt → Polygonkante (alle Ringe).
    public static func distToPolygonEdgeM(_ p: CLLocationCoordinate2D, _ polygon: GZPolygon) -> Double {
        var minimum = Double.infinity
        let cosLat = cos(p.latitude * deg)
        for ring in polygon {
            let count = ring.count / 2
            guard count >= 2 else { continue }
            // Wie v1: bis `count - 1`, der schliessende Doppelpunkt der MVT-Ringe
            // liefert das letzte Segment.
            for i in 0..<(count - 1) {
                let d = distToSegmentM(px: p.longitude, py: p.latitude, cosLat: cosLat,
                                       ax0: ring[2 * i], ay0: ring[2 * i + 1],
                                       bx0: ring[2 * i + 2], by0: ring[2 * i + 3])
                if d < minimum { minimum = d }
            }
        }
        return minimum
    }
}
